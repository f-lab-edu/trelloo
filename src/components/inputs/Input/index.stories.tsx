import React from "react";
import Component from "./index";

export default {
  title: "common/inputs/Input",
  component: Component,
};

const Template = () => {
  const handleSubmit = () => {};

  return <Component onSubmit={handleSubmit} placeHolder="Enter new keyword..." name="input" />;
};

export const Input = Template.bind({});
