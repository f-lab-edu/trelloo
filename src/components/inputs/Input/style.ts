import styled from "styled-components";
import { Input as AntdInput } from "antd";

export const Input = styled(AntdInput).attrs(() => ({}))`
  width: 300px;
`;
