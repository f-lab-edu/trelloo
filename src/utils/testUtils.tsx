import { store } from "@/store";
import { GlobalStyle } from "@/styles/GlobalStyle";
import { theme } from "@/styles/theme";
import ModalsProvider from "@components/modals/ModalsProvider";
import { QueryClient, QueryClientProvider, UseQueryResult } from "@tanstack/react-query";
import { render, renderHook, RenderOptions } from "@testing-library/react";
import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";

const TestProviders = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient();

  return (
    <QueryClientProvider client={queryClient}>
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <GlobalStyle />
          <ModalsProvider>{children}</ModalsProvider>
        </ThemeProvider>
      </Provider>
    </QueryClientProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: TestProviders, ...options });

const createQueryClientWrapper = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient();

  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};

export const createMockedQuery = (hook: () => UseQueryResult<any, unknown>) => {
  return renderHook(() => hook(), { wrapper: createQueryClientWrapper });
};

export * from "@testing-library/react";
export { customRender };
