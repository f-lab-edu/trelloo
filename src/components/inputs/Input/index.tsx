import React from "react";
import { useController, useForm } from "react-hook-form";
import * as S from "./style";

export interface Props {
  onSubmit: (params: Record<string, string>) => void;
  name: string;
  defaultValue: string;
  placeHolder: string;
  onChange?: (data: string) => void;
}

function Input({ placeHolder, name, defaultValue = "", onSubmit, onChange }: Props) {
  const { handleSubmit, control } = useForm({
    defaultValues: {
      [name]: defaultValue,
    },
    mode: "onChange",
  });

  const { field } = useController({
    control,
    name,
  });

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <S.Input
        {...field}
        placeholder={placeHolder}
        onChange={(e) => {
          field.onChange(e.target.value);
          onChange?.(e.target.value);
        }}
      />
    </form>
  );
}

export default Input;
