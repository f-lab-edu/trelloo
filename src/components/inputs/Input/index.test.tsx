import React from "react";
import { customRender, fireEvent } from "@utils/testUtils";
import { describe, vi, expect } from "vitest";
import "@testing-library/jest-dom";
import { act, cleanup } from "@testing-library/react";
import Input from ".";

describe("CardComposer 테스트", () => {
  afterEach(cleanup);
  const handleSubmit = vi.fn();
  const handleChange = vi.fn();

  const setup = () => {
    const { getByPlaceholderText } = customRender(
      <Input
        placeHolder="Enter a keyword..."
        name="filterMenu"
        onSubmit={handleSubmit}
        onChange={handleChange}
        defaultValue=""
      />,
    );
    const input = getByPlaceholderText("Enter a keyword...");

    return { input };
  };

  it("input 값 입력 후 제출하면 onSubmit 호출됨", async () => {
    const { input } = setup();

    act(() => {
      fireEvent.change(input, { target: { value: "title" } });
      fireEvent.submit(input);
    });

    expect(handleSubmit).toHaveBeenCalled();
  });
});
