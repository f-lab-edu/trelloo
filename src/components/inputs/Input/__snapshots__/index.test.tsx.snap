// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardComposer 테스트 > 스냅샷 테스트 1`] = `
{
  "input": <input
    class="ant-input css-dev-only-do-not-override-ed5zg0 sc-gueYoa dfCOkd"
    name="filterMenu"
    placeholder="Enter a keyword..."
    type="text"
    value=""
  />,
}
`;
